src/librustc/ty/layout.rs

```diff
@@ -706,13 +707,34 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                         } else {
                             align = align.max(field.align);
                         }
+
+                        if field.abi != Abi::Uninhabited {
+                            // Normalize scalar_unit to the maximal valid range
+                            let field_abi = match &field.abi {
+                                Abi::Scalar(x) => Abi::Scalar(scalar_unit(x.value)),
+                                Abi::ScalarPair(x, y) => {
+                                    Abi::ScalarPair(
+                                        scalar_unit(x.value),
+                                        scalar_unit(y.value),
+                                    )
+                                }
+                                _ => field.abi.clone(),
+                            };
+
+                            if abi == Abi::Uninhabited || size == Size::ZERO {
```

Comment:
Why is this checking for the size of the entire thing?

Reply:
Ah, size seems to be the size of the largest field so far. That's worth a comment.
I also do not understand why we need two conditions here, shouldn't this simply be "if we have not set an abi yet"? So, shouldn't the `Uninhabited` test be enough?

Generally, all this new code needs a comment explaining its purpose.

Reply:
>  I think we can remove the Uninhabited test, since no fields means size is 0 anyway.

Is  `(u32,  UninhabitedEnum)`  Uninhabited?   What's its size?

Reply:
It is uninhabited at has size 4. Also see https://github.com/rust-lang/rust/issues/49298.